src/ui/reducers/timeline.ts

```diff
@@ -84,6 +91,32 @@ export const {
   pointsReceived,
 } = timelineSlice.actions;
 
+export const pointsReceivedThunk = (points: TimeStampedPoint[]): UIThunkAction => {
+  return (dispatch, getState) => {
+    const state = getState() as UIState;
+    dispatch(pointsReceived(points));
+    const focusRegion = getFocusRegion(state) as UnsafeFocusRegion | null;
+    if (!focusRegion) {
+      return;
+    }
+    const betterFocusStart = points.find(
+      p => BigInt(p.point) > BigInt(focusRegion.begin.point) && p.time < focusRegion.beginTime
```

Comment:
this violates the monotonic nature of points and time which is super weird

Reply:
I...think...this should be...
```js
p => BigInt(p.point) < BigInt(focusRegion.begin.point) && p.time >= focusRegion.beginTime
```

Reply:
Prior to seeing @jcmorrow's diagram, I thought this code was looking to _expand_ the focus region slightly– to include points that are within the user-defined _time_ but outside of the previous points boundaries we knew about.

I now understand this code to be looking to narrow down the points of the focus window itself so that it more tightly fits the user's specified time– so the logic check makes sense  (although it could use an inline comment about its purpose I think).

That being said, should we also add code for the case I mentioned above? Let's say the user selected a range of 500-1,000 ms, and we _had_ a local execution point for 500 ms– but later we happened to get another _earlier point_ that was also at 500ms. I guess this wouldn't happen in this reducer, since we scoped our analysis by the later cutoff point? (This stuff feels confusion-prone, given that we operate with execution points while presenting _time_.)

Reply:
@bvaughn I am not sure I understand the second part? Like, if the user has said they want to load starting at millisecond 550.5 and our nearest point is 550.1, but we find a new point at 550.4... then we should use it? The window can get arbitrarily small, but as long as the relationship between point and time is a monotonic function the logic is always correct.

Reply:
I suppose there could be some kind of floating point imprecision at some level of arbitrary representation, but if we are in that far, then I don't think changing the focus window should matter? The one exception to this might be if the user has specified a *point* they want to use, and because of a combination of the refinement logic + floating point imprecision we move their point. In general, we should probably have a "the user asked for this point" flag, so that we don't do these gymnastics.

Reply:
> Like, if the user has said they want to load starting at millisecond 550.5 and our nearest point is 550.1, but we find a new point at 550.4

No. Different ms values would be handled by the client already, since we have guards to make sure the range is _at least as big as_ the one you defined.

I'm asking about ... imprecision of time stamps. I guess it's not super common, but multiple calls to get system time can return the same value– even though those calls have different execution points. I'm asking about _that_ case.

Reply:
I've added an epsilon for this case.

Reply:
TBH I don't understand how that fixes this.

Reply:
> I'm asking about ... imprecision of time stamps. I guess it's not super common, but multiple calls to get system time can return the same value– even though those calls have different execution points. I'm asking about that case.

If I ask for the system time and I get back `1,000` and then in a few nanoseconds I ask again and I get `1,000` again, then it seems fine to include both of those points in any window that includes either of them? We are trying to fit sub-millisecond precision to a draggable UI, which even on a short replay probably can't be any more precise than a couple of milliseconds, so I don't see what the harm is in choosing the "wrong" nanosecond point to bound that imprecise window by? Besides, this is the case where not only did that weird thing with the system clock happen, but also that was the *exact* time we registered the mouse as picking when dragging around the focus window. Maybe I'm missing some further scenario?

Reply:
But to explain the epsilon thinking further, consider this scenario (I'll annotate points as `ms:point` for brevity:

1. The user indicates with their mouse that they want their window to end at `100.5`.
2. We don't have a point at `100.5`, but we have one at `110.1:123`, so we use that as our point.
3. The user runs an analysis and we end up finding a new point at `101.099:120`. I think we all agree that we should use that point as our new endpoint.
4. The user runs a further analysis, and we find a new point at `100.500099:117`. What do we do? Well, that number is A) a better fit for the window and B) significantly different from our current choice. Therefore, we should use that as our new fit. We are now within epsilon of the requested window.
5. Another points comes in, this time at `100.500095:118`. Now, notice, our `point` part actually went *up*, even though our time value is still technically better. Now we might end up *expanding* the focus window, and later, if another of these cases happens, we could end up shrinking it further again, causing something a little like z-fighting in graphics (https://en.wikipedia.org/wiki/Z-fighting). If we calculate the amount that our new window would be better by we get  `100.500099 - 100.500095`, or `0.000003999999989900971`. That value is smaller than epsilon. We can not really any longer say whether something is a better fit, because we're so close that we are approaching floating point precision (not really, actually, I think we're still several orders of magnitude off, but we are *definitely* approaching the limits of how precise a human hand can drag a mouse, and anyhow we definitely *could* reach the limits to floating point precision, so we might as well call it here rather than at `0.000001`).

Another way to think of epsilon is "What if we multiplied the number by `1/epsilon` and then `Math.floor`ed it so we could just compute in integers?".

Anyways, it sounds like the case that you are thinking of (the system clock gives the same time number for multiple points) would be fine to just leave both points in the window, and this would also accomplish that.